### PR TITLE
Rewrite `RingBuffer`

### PR DESCRIPTION
**Pull request by torkleyy**
This change is [Reviewable](https://reviewable.io/reviews/rustgd/shrev-rs/22)

### CODE REVIEW

File: src/storage.rs [hunk 1]
```diff
@@ -90,207 +242,180 @@ impl<T: 'static> RingBufferStorage<T> {
...
-                    if num_written > max_written {
-                        (i, num_written)
-                    } else {
-                        (index, max_written)
-                    }
-                });
-            self.current_longest_reader = longest_reader;
-            num_written > self.data.len()
+
+        println!(
```
Comment (Rhuagh) [inline src/storage.rs]: We should pull in log, and use that for this, but it at debug or trace level.

File: src/storage.rs [hunk 2]
```diff
@@ -90,207 +242,180 @@ impl<T: 'static> RingBufferStorage<T> {
...
-            reader_id.buffer_id == self.buffer_id,
-            "ReaderID used with an event buffer it's not registered to.  Not permitted!"
-        );
-        let written = unsafe { (*self.reader_internal[reader_id.reader_id].get()).written };
-        let num_written = if self.written < written {
-            self.written + (self.reset_written - written)
-        } else {
-            self.written - written
+        let (last_read_index, gen) = {
+            let mut meta = self.meta.lock();
```
Comment (Rhuagh) [inline src/storage.rs:425]: Is this necessary, since we can guarantee noone's writing when a read occurs?

Comment (torkleyy) [reply src/storage.rs:425]: Yeah well some other reader could access the meta data as well (and write to it). I'm aware we can use `UnsafeCell`, but that would require to add more checks (like does this id belong to this buffer).

**Comment by torkleyy** [comment]
The logic should be ready for review. Now I only need to fix the TODOs and clean up comments and printlns.

**Comment by kvark** [comment]
those commits :rofl: 
> Work
> Work
> ...
> Nearly there

**Comment by torkleyy** [comment]
Yay, ready for review!

File: src/lib.rs [hunk 3]
```diff
@@ -3,12 +3,16 @@
 //!
 //! See examples directory for examples.
 
-#![deny(missing_docs)]
+#![warn(missing_docs)]
```
Comment (AnneKitsune) [inline src/lib.rs:6]: If you are going for 1.0 you might want to keep deny.

File: src/lib.rs [hunk 4]
```diff
@@ -99,14 +104,41 @@ where
...
+            data,
+            vec![9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]
+        );
+
+        for i in 23..10_000 {
+            channel.single_write(i);
+        }
+
+        let data = channel.read(&mut reader1).cloned().collect::<Vec<_>>();
+        assert_eq!(data, (1..10_000).collect::<Vec<_>>());
```
Comment (AnneKitsune) [inline src/lib.rs:139]: A #[bench] could be nice too ;)

File: src/storage.rs [hunk 5]
```diff
@@ -1,296 +1,494 @@
...
+    /// the end.
+    fn magic(size: usize) -> Self {
+        CircularIndex::new(!0, size)
+    }
+
+    fn is_magic(&self) -> bool {
+        self.index == !0
+    }
+
+    fn step(&mut self, inclusive_end: usize) -> Option<usize> {
```
Comment (AnneKitsune) [inline src/storage.rs:40]: I'm not a big fan of using magic like !0 to indicate the end of the steps. If inclusive_end == !0 you'll skip an event.

File: src/storage.rs [hunk 6]
```diff
@@ -1,296 +1,494 @@
...
 }
 
-impl<T> Drop for ReaderId<T> {
-    fn drop(&mut self) {
-        self.alive.store(false, Ordering::Relaxed);
+impl Add<usize> for CircularIndex {
+    type Output = usize;
+
+    fn add(self, rhs: usize) -> usize {
+        (self.index + rhs) % self.size
```
Comment (AnneKitsune) [inline src/storage.rs:63]: wait what? Not sure I understand why that's like that and AddAssign isn't

File: src/storage.rs [hunk 7]
```diff
@@ -1,296 +1,494 @@
...
 
+impl SubAssign<usize> for CircularIndex {
+    fn sub_assign(&mut self, rhs: usize) {
+        self.index = *self - rhs;
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+struct Data<T> {
```
Comment (AnneKitsune) [inline src/storage.rs:89]: I understand where the name comes from, but it doesn't actually says much about its usage.

File: src/storage.rs [hunk 8]
```diff
@@ -1,296 +1,494 @@
...
-struct InternalReaderId {
-    written: usize,
-    index: usize,
-    alive: Arc<AtomicBool>,
+pub struct ReaderId<T: 'static> {
+    id: usize,
+    marker: PhantomData<&'static [T]>,
+    // stupid way to make this `Sync`,
+    // never actually locked
+    drop_notifier: Mutex<Sender<usize>>,
```
Comment (AnneKitsune) [inline src/storage.rs:210]: Comment ^

File: src/storage.rs [hunk 9]
```diff
@@ -1,296 +1,494 @@
...
-{
+/// Ring buffer, holding data of type `T`.
+#[derive(Debug)]
+pub struct RingBuffer<T> {
+    available: usize,
+    last_index: CircularIndex,
+    data: Data<T>,
+    // stupid way to make this `Sync`,
+    // never actually locked
+    free_rx: Mutex<Receiver<usize>>,
```
Comment (AnneKitsune) [inline src/storage.rs:281]: Comment ^

File: src/storage.rs [hunk 10]
```diff
@@ -1,296 +1,494 @@
...
+    // stupid way to make this `Sync`,
+    // never actually locked
+    free_tx: Mutex<Sender<usize>>,
+    generation: Wrapping<usize>,
+    nearest_reader: usize,
+    meta: Mutex<ReaderMeta>, // TODO: consider `UnsafeCell`
```
Comment (AnneKitsune) [inline src/storage.rs:287]: TODO

File: src/storage.rs [hunk 11]
```diff
@@ -1,296 +1,494 @@
...
-            // In order to avoid pushing events that have already been read back into a readable
-            // range we're also going to push any readers that are ahead of us in the buffer
-            // forward one as well.
-            for i in 0..self.reader_internal.len() {
-                unsafe {
-                    if (*self.reader_internal[i].get()).index > self.write_index {
-                        (*self.reader_internal[i].get()).index += 1;
-                    }
-                }
+    #[inline(never)]
```
Comment (AnneKitsune) [inline src/storage.rs:349]: I just read on #inline and I think this is not required and would slow the program? Is it the objective? "ensure_additional_SLOW"

**Review by AnneKitsune — CHANGES_REQUESTED** [review]
Since I saw you were going for 1.0.0, I made a harsher(?) review.

I'm not going to pretend I understand exactly what the change actually does except manually handling the memory in the ring buffer and changing the grow algorithm. Thus, I have a couple of comments that just point at things that might or might not look weird.

File: src/lib.rs (hunk 3 above)
Comment (Xaeroxe) [reply src/lib.rs:6]: I disagree, 1.0 doesn't mean no more development is going to happen it just indicates the code is considered "production ready".

File: src/storage.rs (hunk 8 above)
Comment (Xaeroxe) [reply src/storage.rs:210]: Alternatively you could just `unsafe impl Sync` for this

File: src/lib.rs (hunk 3 above)
Comment (AnneKitsune) [reply src/lib.rs:6]: which is why I think docs are essential for "production ready"

Comment (AnneKitsune) [reply src/lib.rs:6]: Changing to warn during development is fine, but I usually turn it back to deny when pushing code.

**Review by Rhuagh — APPROVED** [review]
LGTM

File: src/storage.rs (hunk 8 above)
Comment (AnneKitsune) [reply src/storage.rs:210]: That's what there was before, I don't know why it was changed.

File: src/lib.rs (hunk 3 above)
Comment (Xaeroxe) [reply src/lib.rs:6]: Warnings should all be resolved by the time a PR is merged, if they aren't then the developers weren't careful enough imo.  Keeping a deny(missing_docs) makes prototyping more difficult and doesn't even work.  Most of the time this causes people to override the compiler by writing

```rust
/// 
pub fn thing()
```

This happened to the amethyst renderer at one point, and leaving that kind of "documentation" in without a warning is a much bigger problem.

Comment (Rhuagh) [reply src/lib.rs:6]: warn is fine, deny is too restrictive imho

File: src/storage.rs (hunk 5 above)
Comment (Rhuagh) [reply src/storage.rs:40]: much better than the alternative (double special case)

File: src/lib.rs (hunk 3 above)
Comment (AnneKitsune) [reply src/lib.rs:6]: okay

File: src/storage.rs (hunk 6 above)
Comment (Rhuagh) [reply src/storage.rs:63]: Because `AddAssign` use this impl internally.

File: src/storage.rs (hunk 11 above)
Comment (torkleyy) [reply src/storage.rs:349]: Not sure what you mean exactly.

File: src/storage.rs (hunk 10 above)
Comment (torkleyy) [reply src/storage.rs:287]: not for this PR

File: src/storage.rs (hunk 9 above)
Comment (torkleyy) [reply src/storage.rs:281]: yeah I think it's a useful hint for somebody who reads the code? It's only as long as we don't have a better, clean solution.

File: src/storage.rs (hunk 8 above)
Comment (torkleyy) [reply src/storage.rs:210]: same as below

File: src/storage.rs (hunk 7 above)
Comment (torkleyy) [reply src/storage.rs:89]: what do you suggest?

File: src/lib.rs (hunk 4 above)
Comment (torkleyy) [reply src/lib.rs:139]: Agreed :)

**Comment by torkleyy** [comment]
@jojolepro I made it `1.0` because I think the API is stable. Not to state it's perfectly documented :smile:

**Comment by AnneKitsune** [comment]
https://stackoverflow.com/questions/37639276/when-should-inline-be-used-in-rust?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa

**Pull request merged**